ktor-network/posix/src/io/ktor/network/sockets/DatagramSocketImpl.kt

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.network.util.*
+import io.ktor.util.network.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.errors.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import platform.posix.*
+import kotlin.coroutines.*
+
+internal class DatagramSocketImpl(
+    private val descriptor: Int,
+    val selector: SelectorManager,
+    private val _localAddress: NetworkAddress,
+    private val _remoteAddress: NetworkAddress?,
+    parent: CoroutineContext = EmptyCoroutineContext
+) : BoundDatagramSocket, ConnectedDatagramSocket, Socket, CoroutineScope {
+    private val _context: CompletableJob = Job(parent[Job])
+    val selectable: SelectableNative = SelectableNative(descriptor)
+
+    override val coroutineContext: CoroutineContext = parent + Dispatchers.Unconfined + _context
+
+    override val socketContext: Job
+        get() = _context
+
+    override val localAddress: NetworkAddress
+        get() = _localAddress
+    override val remoteAddress: NetworkAddress
+        get() = _remoteAddress!! // TODO: What should happen here?
+
+    private val sender: SendChannel<Datagram> = DatagramSendChannel(descriptor, this)
+
+    private val receiver: ReceiveChannel<Datagram> =
+        produce(coroutineContext) { // TODO: Dispatchers.IO which is used in jvm version of DatagramSocketImpl is unavailable
```

Comment:
The JVM version of this file uses the IO dispatcher here, but that is unavailable on Kotlin/Native.

Reply:
and same here